### PR TITLE
[Intel GPU] xpu-ops codegen via backend whitelist

### PR DESCRIPTION
**Pull request by ZhiweiYan-96**
# Motivation

This PR intends to enhance the `gen.py` to allow generate codes for a specific backend.

XPU operators need be registered in an hand-written way currently. Developers have no chance to take the advantage of shared code to handle tensor meta setting (like strides, proxy output, structured kernels).  Manually porting code is erro-prone and may lead to high maintaining efforts.

We intend to enhance the `bckend_whitelist` argument in `gen.py` to generate independent backend specific files. Then backend developers could directly use such generated code

# Usage
The specialty of XPU ops is that they lie in `third_pary/torch-xpu-ops`, so it codegen resides in a separate process.

We use the following commands to trigger XPU codegen

` python -m torchgen.gen --source-path path/to/yaml/of/xpu   --install-dir  build/aten/src/ATen/xpu    --per-operator-headers    --static-dispatch-backend     --backend-whitelist=XPU`

The diff lies at `backend-whitelist=XPU`.  The backend-whitelist key is an existent argument in torchgen. But we observe that it can be hardly used for generating source files like `Registexxx.cpp`. We enhance the backend-white list argument in `gen.py`, where we could use it to generate backend only code. Like `build/aten/src/ATen/xpu/RegisterXPU.cpp, build/aten/src/ATen/xpu/ops/as_strided_native.h`

The input of `gen.py` are code templates and operators yaml. We share the same templates in `aten`. A simplified yaml lies in `third_party/torch-xpu-ops`, which only includes the supported xpu operators. This yaml is a copy-and-modify of `native_functions.yaml`. No extra entry is added, the format is same as the one in `aten`

# Result

All operators headers are generated in `build/aten/src/ATen/xpu/ops` independently, which would not affect operators declared/defined by CPU/CUDA or any other backend.  XPU operators only include headers in this folder.

# Verification
* We add a ut for verifying xpu codgen in `test_codegen.py` and it could be triggered through
`python -m tools.test.test_codegen -k BackendWhitelist`

* In `third-party/torch-xpu-ops`, we migrate some operators to structured kernels style, where they are registered through `REGISTER_XPU_DISPATCH` or `TORCH_IMPL_FUNC`. And it is tested with `test_xpu.py`

# Others
As for the longer term, we would unify code generation system with other backend also, like cuda, mps.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129452
* __->__ #127865
* #127860

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @chauhang @d4l3k @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel @EikanWang @voznesenskym @penguinwu @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @LucasLLC @MeetVadakkanchery @mhorowitz

### CODE REVIEW

File: torchgen/dest/register_dispatch_key.py [hunk 1]
```diff
@@ -58,6 +59,8 @@ def gen_registration_headers(
             headers.append("#include <ATen/cuda/EmptyTensor.h>")
     elif backend_index.dispatch_key == DispatchKey.MPS:
         headers.append("#include <ATen/mps/EmptyTensor.h>")
+    elif backend_index.dispatch_key == DispatchKey.XPU:
+        headers.append("#include <aten/EmptyTensor.h>")
```
Comment (gujinghui) [inline torchgen/dest/register_dispatch_key.py:63]: This file is not built-in pytorch? Should we add more comments?

File: torchgen/gen.py [hunk 2]
```diff
@@ -1955,11 +1980,20 @@ def gen_per_operator_headers(
         ("NativeMetaFunctions", "_meta"),
         ("NativeFunctions", "_native"),
     ]:
+        if (whitelist_keys is not None) and (len(whitelist_keys) == 1):
+            dispatch_namespace = list(whitelist_keys)[0].lower()
+        #     includes = f"#include <ATen/{dispatch_namespace}/ops/{name}{suffix}.h>"
+        # else:
+        #     includes = f"#include <ATen/ops/{name}{suffix}.h>"
```
Comment (gujinghui) [inline torchgen/gen.py:1987]: if not needed, just remove it.

File: torchgen/dest/register_dispatch_key.py (hunk 1 above)
Comment (ZhiweiYan-96) [reply torchgen/dest/register_dispatch_key.py:64]: sure, I have added comment to refer the function and location, thank.

File: torchgen/gen.py (hunk 2 above)
Comment (ZhiweiYan-96) [reply torchgen/gen.py]: has removed, thanks for point out the issue

File: torchgen/gen.py [hunk 3]
```diff
@@ -1451,6 +1464,12 @@ def get_ns_grouped_kernels(
         native_function_namespaces = set()
         dispatch_keys = set()
         for dispatch_key, backend_idx in backend_indices.items():
+            if (
+                whitelist_keys
+                and len(whitelist_keys) > 0
+                and (dispatch_key not in whitelist_keys)
+            ):
```
Comment (EikanWang) [inline torchgen/gen.py:1471]:
```suggestion
            if whitelist_keys and dispatch_key not in whitelist_keys:
```

File: torchgen/gen.py [hunk 4]
```diff
@@ -1955,16 +1980,36 @@ def gen_per_operator_headers(
...
+                    f"{category}.h",
+                    lambda: {
+                        f"{category}_includes": [
+                            f"#include <ATen/{dispatch_namespace}/ops/{name}{suffix}.h>"
+                            for name in sorted(functions_by_root_name.keys())
+                        ],
+                        f"{category}_declarations": [],
+                    },
+                )
+        else:
```
Comment (EikanWang) [inline torchgen/gen.py:2002]: Do we need to handle multiple `white list keys` now? Should we just add an assert here? Like:
```python
assert not whitelist_keys or len(whiltelist_keys) == 1
```

File: torchgen/gen.py [hunk 5]
```diff
@@ -2245,16 +2293,30 @@ def operator_headers() -> List[str]:
                     if not is_registered:
                         continue
 
-                    headers.append(f"#include <ATen/ops/{g.root_name}_native.h>")
+                    if (whitelist_keys is not None) and (
+                        dispatch_key in whitelist_keys
+                    ):
```
Comment (EikanWang) [inline torchgen/gen.py:2298]:
```suggestion
                    if whitelist_keys and dispatch_key in whitelist_keys:
```

File: torchgen/gen.py [hunk 6]
```diff
@@ -2245,16 +2293,30 @@ def operator_headers() -> List[str]:
...
                         == DispatchKey.CompositeExplicitAutogradNonFunctional
                     ):
                         headers.append(f"#include <ATen/ops/{g.root_name}.h>")
                     if dispatch_key in functions_keys:
-                        headers.append(
-                            f"#include <ATen/ops/{g.root_name}_{dispatch_namespace}_dispatch.h>"
-                        )
+                        if (
+                            whitelist_keys is not None
+                        ) and dispatch_key in whitelist_keys:
```
Comment (EikanWang) [inline torchgen/gen.py:2312]: ditto

File: torchgen/gen.py [hunk 7]
```diff
@@ -2450,6 +2516,11 @@ def headers_for_aoti() -> str:
 
         del fm
 
+    if (whitelist_keys is not None) and (
+        not len(whitelist_keys) == 0
+    ):  # Only generate backend required source files
```
Comment (EikanWang) [inline torchgen/gen.py:2521]:
```suggestion
    if whitelist_keys:  # Only generate backend required source files
```

File: torchgen/gen.py [hunk 8]
```diff
@@ -2915,7 +3001,7 @@ def main() -> None:
         dispatch_keys = [
             k
             for k in dispatch_keys
-            if is_generic_dispatch_key(k) or str(k) in options.backend_whitelist
+            if (is_generic_dispatch_key(k) or str(k) in options.backend_whitelist)
```
Comment (EikanWang) [inline torchgen/gen.py:3004]:
```suggestion
            if is_generic_dispatch_key(k) or str(k) in options.backend_whitelist
```

File: torchgen/gen.py [hunk 9]
```diff
@@ -2825,6 +2896,11 @@ def main() -> None:
         default=["headers", "sources", "declarations_yaml"],
         help="Generate only a subset of files",
     )
+    parser.add_argument(
+        "--only_backend",
```
Comment (EikanWang) [inline torchgen/gen.py:2900]: Is this option useful to produce code for another backend?

**Review by EikanWang — CHANGES_REQUESTED** [review]
Please add UT.

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: torchgen/gen.py [hunk 10]
```diff
@@ -2853,6 +2929,15 @@ def main() -> None:
             del dispatch_keys[dispatch_keys.index(DispatchKey.MPS)]
 
     parsed_yaml = parse_native_yaml(native_yaml_path, tags_yaml_path, ignore_keys)
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply torchgen/gen.py]: has removed, thanks for reminding

File: torchgen/dest/register_dispatch_key.py [hunk 11]
```diff
@@ -43,6 +43,7 @@ def gen_registration_headers(
     backend_index: BackendIndex,
     per_operator_headers: bool,
     rocm: bool,
+    backend_only: bool = False,
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply torchgen/dest/register_dispatch_key.py]: has removed, thanks

File: torchgen/gen.py (hunk 7 above)
Comment (ZhiweiYan-96) [reply torchgen/gen.py]: modified

File: torchgen/gen.py (hunk 8 above)
Comment (ZhiweiYan-96) [reply torchgen/gen.py]: modifed

File: torchgen/gen.py (hunk 9 above)
Comment (ZhiweiYan-96) [reply torchgen/gen.py]: removed

File: torchgen/dest/register_dispatch_key.py [hunk 12]
```diff
@@ -58,6 +59,9 @@ def gen_registration_headers(
             headers.append("#include <ATen/cuda/EmptyTensor.h>")
     elif backend_index.dispatch_key == DispatchKey.MPS:
         headers.append("#include <ATen/mps/EmptyTensor.h>")
+    elif backend_index.dispatch_key == DispatchKey.XPU:
+        # XPU specific, this header resides in third_party/torch-xpu-ops
+        headers.append("#include <aten/EmptyTensor.h>")
```
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply torchgen/dest/register_dispatch_key.py:63]: Then I may need move all source files in `torch_xpu_ops/src/aten/` into `torch_xpu_ops/src/aten/xpu`.
Does these changes have any risk @fengyuan14 for us?

[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply torchgen/dest/register_dispatch_key.py:63]: synced with @fengyuan14 , should be no big change for torch-xpu-ops, I am WIP

**Comment by EikanWang** [comment]
@ZhiweiYan-96 , please fix UT failures and we need to enable it ASAP.

File: aten/src/ATen/xpu/xpu_functions.yaml [hunk 13]
```diff
@@ -0,0 +1,55 @@
+# This yaml is used only for testing the XPU codegen functionality.
```
Comment (EikanWang) [inline aten/src/ATen/xpu/xpu_functions.yaml:1]: Why do we need to add this yaml to the source code? I think it should be the data file of test folder. Why is it not be a part of `test/xpu`?

**Review by EikanWang — CHANGES_REQUESTED** [review]

File: aten/src/ATen/xpu/xpu_functions.yaml (hunk 13 above)
Comment (ZhiweiYan-96) [reply aten/src/ATen/xpu/xpu_functions.yaml]: Indeed, have moved into test/xpu, thanks for the comment

File: tools/test/test_gen_backendwhitelist.py
[the remark this answers is not in the document]
Comment (ZhiweiYan-96) [reply tools/test/test_gen_backendwhitelist.py:1]: Yes, I have moved the ut into `test_codegen.py`

**Pull request closed without merging**